### PR TITLE
Document and test building for older Kotlin language versions

### PR DESCRIPTION
**Pull request by eskatos**
### Context

* Fixes https://github.com/gradle/gradle/issues/25868 by documenting the limitation in the upgrading guide
* Adds integration tests setting support boundaries and demonstrating a workaround to build for previous Kotlin versions with the current Gradle.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test QF BD

File: subprojects/kotlin-dsl-plugins/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt [hunk 1]
```diff
@@ -71,4 +66,108 @@ class KotlinDslPluginCrossVersionSmokeTest : AbstractKotlinIntegrationTest() {
...
+        }
+    }
+
+    @Test
+    fun `can build plugin for previous unsupported Kotlin language version`() {
+
+        // Kotlin version leaks on the classpath when running embedded
+        assumeNonEmbeddedGradleExecuter()
+
+        val previousKotlinLanguageVersion = "1.2"
```
Comment (ljacomet) [inline subprojects/kotlin-dsl-plugins/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt:107]: 🤔 Does it make sense to use a Kotlin version that Gradle never really used? Should we stick with 1.3 instead?

**Review by ljacomet — APPROVED** [review]

File: subprojects/kotlin-dsl-plugins/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt (hunk 1 above)
Comment (eskatos) [reply subprojects/kotlin-dsl-plugins/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginCrossVersionSmokeTest.kt:107]: I captured the oldest we can use. I think it has value given before Gradle 5.0 we already had an older embedded Kotlin, it's just not advertised on the compatibility matrix.

**Comment by eskatos** [comment]
@bot-gradle test and merge

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 2]
```diff
@@ -93,6 +93,9 @@ The embedded Kotlin has been updated from 1.8.20 to link:https://github.com/JetB
 The Kotlin language and API levels for the Kotlin DSL are still set to 1.8 for backwards compatibility.
 See the release notes for link:https://github.com/JetBrains/kotlin/releases/tag/v1.8.22[Kotlin 1.8.22] and link:https://github.com/JetBrains/kotlin/releases/tag/v1.8.21[Kotlin 1.8.21].
 
+Kotlin 1.9 dropped support for Kotlin language and API level 1.3.
+If you build Gradle plugins written in Kotlin with this version of Gradle and need to support Gradle <7.0 you need to stick to using the Kotlin Gradle Plugin <1.9.0 and configure the Kotlin language and API levels to 1.3.
```
Comment (TWiStErRob) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:97]:
```suggestion
If you build Gradle plugins written in Kotlin with this version of Gradle and need to support Gradle <7.0 you need to stick to using the Kotlin Gradle Plugin <1.9.0 and configure the Kotlin language and API levels to 1.3. When using the `kotlin-dsl` plugin, you need to use version "4.0.14" or lower.
```

Comment (eskatos) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:97]: As explained in https://github.com/gradle/gradle/issues/25868#issuecomment-1650063881, the workaround is unsupported, using previous versions of the `kotlin-dsl` plugin is also unsupported. That's why I omitted this from the upgrading guide on purpose.

Comment (TWiStErRob) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:97]: Makes sense!

**Review by lkasso — APPROVED** [review]

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Pull request merged**